### DISCUSSION TITLE
RFC (request for comments): CAN bus plugin

### DISCUSSION THREAD

**Discussion by dresco** [grblHAL/core#196] [Show and tell]
I've started working on a CAN bus plugin, and wanted to share initial progress to get some feedback. (My use case for this is wanting to update my [Control Panel](https://github.com/dresco/Plugin_panel) plugin to use CAN bus as an option instead of MODBUS).

Here are the links to the code I have so far;
- The CAN bus [plugin](https://github.com/dresco/Plugin_canbus) itself
- A minor [core](https://github.com/dresco/core/commit/4a25254bd73a33393e54ae9c3f336c6d3a41d8ad) update to support settings registration
- The supporting lower level CAN driver code for [STM32F4xx](https://github.com/dresco/STM32F4xx/commit/d1340f6ddf52a21e87676bfcfbef7afd91a2fb2d) driver

If in agreement that this approach is okay, I next plan to add lower level drivers for more platforms. Probably F7, H7, iMXRT1062, and the Microchip MCP2510 SPI CAN controller (for platforms without a CAN peripheral).

**Comment by terjeio** [comment]
Ok for me, the only thing I can see is missing is definition of a group name for `Group_VFD` in the plugin.

A minor thing is that [MCP2510](https://www.microchip.com/en-us/product/MCP2510) (and MCP2515 that I have worked with before) is not recommended for new designs, Microchip writes that [MCP2518FP](https://www.microchip.com/en-us/product/mcp2518fd) should be used instead. I have not checked to see if these chips have the same SPI interface protocol.

**Reply by dresco** [reply]
Ah yes, I see the bits I'd missed for `Group_CANbus` - will add those.

Thanks for the heads up on the MCP2510, I've not ordered one yet, so will pick up the new part instead. 

From memory I'm not sure the existing Teensy boards expose any CAN or SPI pins to headers? But I guess if the software support is there, then that's an incentive for future boards..

**Reply by dresco** [reply]
@terjeio Just checking before I commit, I think these `setting_group_detail_t` lines are what I'm missing for the settings description?

```
diff --git a/canbus.c b/canbus.c
index 477b21d..5248899 100644
--- a/canbus.c
+++ b/canbus.c
@@ -81,11 +81,17 @@ uint8_t canbus_queue_rx(canbus_message_t message)
     return (queue_put_msg(&rx_buffer, message));
 }
 
+static const setting_group_detail_t canbus_groups [] = {
+    { Group_Root, Group_CANbus, "CAN bus"}
+};
+
 static const setting_detail_t canbus_setting_detail[] = {
     { Setting_CANbus_BaudRate, Group_CANbus, "CAN bus baud rate", NULL, Format_RadioButtons, "125000,250000,500000,1000000", NULL, NULL, Setting_NonCoreFn, canbus_set_baud, canbus_get_baud, NULL },
 };
 
 static setting_details_t setting_details = {
+    .groups = canbus_groups,
+    .n_groups = sizeof(canbus_groups) / sizeof(setting_group_detail_t),
     .settings = canbus_setting_detail,
     .n_settings = sizeof(canbus_setting_detail) / sizeof(setting_detail_t),
     .save = canbus_settings_save,
```

**Reply by terjeio** [reply]
Seems correct to me. If shows up in ioSender as it should then sure.

**Reply by dresco** [reply]
> Seems correct to me. If shows up in ioSender as it should then sure.

Cool thanks, I'll submit a PR for the core settings change - to get the baud rate setting id reserved..
